File: src/me/krickl/memebotj/UserHandler.scala
package me.krickl.memebotj

import java.math.BigInteger
import java.security.SecureRandom
import java.text.SimpleDateFormat
import java.util
import java.util.Calendar
import java.util.logging.Logger

import com.mongodb.client.{FindIterable, MongoCollection}
import me.krickl.memebotj.Services.Twitch.UserAPI
import me.krickl.memebotj.Utility.{CommandPower, Cooldown}
import org.bson.Document

import scala.beans.BeanProperty

//remove if not needed
import scala.collection.JavaConversions._

object UserHandler {
  final var log = Logger.getLogger(UserHandler.getClass.getName)
}

/**
  * This class handles users
  *
  * @author unlink
  *
  */
class UserHandler(@BeanProperty var username: String, @BeanProperty var channelOrigin: String) {
  @BeanProperty
  var isModerator: Boolean = false
  @BeanProperty
  var isUserBroadcaster: Boolean = false
  //private boolean execCommands = true
  @BeanProperty
  var newUser: Boolean = false
  @BeanProperty
  var nickname = ""
  var _commandPower: Int = CommandPower.viewerAbsolute
  var _autoCommandPower: Int = CommandPower.viewerAbsolute
  @BeanProperty
  var customCommandPower: Int = 0
  private var _points: Double = 0.0f

  def points = _points

  // private boolean isJoined = false
  @BeanProperty
  var userCooldown = new Cooldown(0)
  @BeanProperty
  var autogreet: String = ""
  @BeanProperty
  var timeouts: Int = 0
  @BeanProperty
  var userCollection: MongoCollection[Document] = null
  @BeanProperty
  var userCommandCooldowns: java.util.HashMap[String, Cooldown] = new java.util.HashMap[String, Cooldown]()
  @BeanProperty
  var modNote: String = ""
  @BeanProperty
  var random = new SecureRandom()
  @BeanProperty
  var privateKey = new BigInteger(130, random).toString(32)

  @BeanProperty
  var dateJoined = "null"
  @BeanProperty
  var timeStampJoined = System.currentTimeMillis()

  @BeanProperty
  var enableAutogreets = true

  //this will be used for spam prevention
  @BeanProperty
  var lastMessages = Array.ofDim[String](10)

  var timeSinceActivity = System.currentTimeMillis()

  var walletSize: Double = -1

  var isFollowing = false
  var hasFollowed = false

  var hasAutogreeted = false

  var removeCooldown: Cooldown = new Cooldown(0)
  var shouldBeRemoved = false

  val twitchUserAPI = new UserAPI(this.username)

  if (Memebot.useMongo) {
    this.userCollection = Memebot.db.getCollection(this.channelOrigin + "_users")
  }

  readDBUserData()

  if (dateJoined == "null") {
    val sdfDate = new SimpleDateFormat("yyyy-MM-dd hh:mm:ss a") // dd/MM/yyyy
    val cal = Calendar.getInstance()
    this.dateJoined = sdfDate.format(cal.getTime)
  }

  setCommandPower(this._autoCommandPower)

  UserHandler.log.info(String.format("Private key for user %s is %s", this.username, this.privateKey))

  def writeDBUserData(): Unit = {
    if (!Memebot.useMongo) {
      return
    }

    val channelQuery = new Document("_id", this.username)

    val channelData = new Document("_id", this.username).append("pointsf", this._points)
      .append("mod", this.isModerator).append("autogreet", this.autogreet)
      .append("ccommandpower", this.customCommandPower).append("broadcaster", this.isUserBroadcaster)
      .append("timeouts", this.timeouts)
      .append("privatekey", this.privateKey)
      .append("enableautogreet", this.enableAutogreets)
      .append("datejoined", this.dateJoined)
      .append("timeStampJoined", this.timeStampJoined)
      .append("nickname", this.nickname)
      .append("wallet", this.walletSize)
      .append("isfollowing", this.isFollowing)
      .append("hasfollowed", this.hasFollowed)

    try {
      if (this.userCollection.findOneAndReplace(channelQuery, channelData) == null) {
        this.userCollection.insertOne(channelData)
      }
    } catch {
      case e: Exception =>
        e.printStackTrace()
    }
  }

  def readDBUserData(): Unit = {
    if (!Memebot.useMongo) {
      return
    }

    val channelQuery = new Document("_id", this.username)
    val cursor: FindIterable[Document] = this.userCollection.find(channelQuery)

    val channelData = cursor.first()

    // read data
    if (channelData != null) {
      this.isModerator = channelData.getBoolean("mod", this.isModerator)
      this._points = channelData.getOrDefault("pointsf", this._points.asInstanceOf[Object]).asInstanceOf[Double]
      this.autogreet = channelData.getOrDefault("autogreet", this.autogreet).toString
      this.customCommandPower = channelData.getOrDefault("ccommandpower", this.customCommandPower.asInstanceOf[Object]).asInstanceOf[Int]
      this.isUserBroadcaster = channelData.getOrDefault("broadcaster", this.isUserBroadcaster.asInstanceOf[Object]).asInstanceOf[Boolean]
      this.timeouts = channelData.getOrDefault("timeouts", this.timeouts.asInstanceOf[Object]).asInstanceOf[Int]
      this.privateKey = channelData.getOrDefault("privatekey", this.privateKey.asInstanceOf[Object]).asInstanceOf[String]
      this.enableAutogreets = channelData.getOrDefault("enableautogreet", this.enableAutogreets.toString).toString.toBoolean
      this.dateJoined = channelData.getOrDefault("datejoined", this.dateJoined).toString
      this.timeStampJoined = channelData.getOrDefault("timeStampJoined", this.timeStampJoined.asInstanceOf[Object]).asInstanceOf[Long]
      this.nickname = channelData.getOrDefault("nickname", this.nickname.asInstanceOf[Object]).asInstanceOf[String]
      this.walletSize = channelData.getOrDefault("wallet", this.walletSize.asInstanceOf[Object]).asInstanceOf[Double]
      this.isFollowing = channelData.getOrDefault("isfollowing", this.isFollowing.asInstanceOf[Object]).asInstanceOf[Boolean]
      this.hasFollowed = channelData.getOrDefault("hasfollowed", this.hasFollowed.asInstanceOf[Object]).asInstanceOf[Boolean]
    } else {
      this.newUser = true
    }
  }

  def update(channelHandler: ChannelHandler = null) = {
    twitchUserAPI.update(this, channelHandler)
  }

  def sendAutogreet(channelHandler: ChannelHandler): Unit = {
    if (!this.hasAutogreeted && this.enableAutogreets && channelHandler.allowAutogreet && this.autogreet != "") {
      channelHandler.sendMessage(this.autogreet, sender = this)
      this.hasAutogreeted = true
    }
  }

  /**
    * @deprecated compatibility for old function calls. This function should not be used anymore.
    * @param f new amount
    * @return
    */
  @Deprecated
  def setPoints(f: Double): Boolean = {
    this.points = f
  }

  def points_=(f: Double): Boolean = {
    var result = true

    this._points = f
    if (this._points < 0) {
      this._points = 0
    }

    for (ch <- Memebot.joinedChannels) {
      if (ch.channel == this.channelOrigin) {
        if (this._points + f > ch.maxPoints || (this._points + f > this.walletSize && this.walletSize > 0)) {
          this._points = ch.maxPoints
          result = false
        }
      }
    }

    result
  }

  @Deprecated
  def setCommandPower(commandPower: Int) = {
    this.commandPower_=(commandPower)
  }

  def commandPower_=(commandPower: Int) = {
    this._autoCommandPower = commandPower
    this._commandPower = commandPower + this.customCommandPower
  }

  @Deprecated
  def setAutoCommandPower(autoCommandPower: Int) = {
    this.autoCommandPower_=(autoCommandPower)
  }

  def autoCommandPower_=(autoCommandPower: Int) = {
    this._autoCommandPower = autoCommandPower
  }

  def screenName: String = {
    if (!this.nickname.isEmpty && Memebot.debug) {
      return nickname
    }
    this.username
  }

  def canRemove: Boolean = {
    // remove user after 1 hour (0x36EE80 milliseconds) of inactivity
    val tmpUserList = new util.HashMap[String, UserHandler]()
    tmpUserList.put(this.username, this)
    if ((System.currentTimeMillis() - this.timeSinceActivity) > 0x36EE80 * 0x2 && username != "#internal#" && username != "#readonly#" && !CommandHandler.checkPermission(this, CommandPower.broadcasterAbsolute, tmpUserList)) {
      UserHandler.log.info(f"Removed user ${this.username} for inactivity ${this.timeSinceActivity}")
      return true
    }

    //check if user has been marked for removal
    if (this.shouldBeRemoved && this.removeCooldown.canContinue) {
      UserHandler.log.info(f"Removed user ${this.username}")
      return true
    }

    false
  }
}

File: src/main/java/me/krickl/memebotj/UserHandler.java
package me.krickl.memebotj;

import me.krickl.memebotj.Database.MongoHandler;
import me.krickl.memebotj.Exceptions.DatabaseReadException;
import me.krickl.memebotj.Inventory.Inventory;
import me.krickl.memebotj.Utility.CommandPower;
import me.krickl.memebotj.Utility.Cooldown;
import org.apache.commons.codec.binary.Base64;
import org.bson.Document;
import org.json.simple.JSONObject;

import java.security.SecureRandom;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.HashMap;
import java.util.logging.Logger;

/**
 * This file is part of memebotj.
 * Created by unlink on 06/04/16.
 */
public class UserHandler implements Comparable<UserHandler> {
    public static Logger log = Logger.getLogger(UserHandler.class.getName());
    boolean isModerator = false;
    boolean isUserBroadcaster = false;
    boolean newUser = false;
    int commandPower = CommandPower.viewerAbsolute;
    int autoCommandPower = CommandPower.viewerAbsolute;
    int customCommandPower = 0;
    int timeouts = 0;
    boolean shouldBeRemoved = false;
    private MongoHandler mongoHandler = null;
    private String username = "";
    private String channelOrigin = "";
    private String nickname = "";
    private double points = 0.0f;
    // private boolean isJoined = false
    private Cooldown userCooldown = new Cooldown(0);
    private String autogreet = "";
    // todo old db code - remove soon
    //MongoCollection<Document> userCollection = null;
    private java.util.HashMap<String, Cooldown> userCommandCooldowns = new java.util.HashMap<String, Cooldown>();
    private String modNote = "";
    private SecureRandom random = new SecureRandom();
    private String dateJoined = "null";
    private long timeStampJoined = System.currentTimeMillis();
    private boolean enableAutogreets = true;
    private double walletSize = -1;
    private boolean isFollowing = false;
    private boolean hasFollowed = false;
    private boolean hasAutogreeted = false;
    private Cooldown removeCooldown = new Cooldown(0);
    private int jackpotWins = 0;

    private Inventory userInventory;

    private int lastTimeoutDuration = 0;
    private String lastTimeoutReason = "";

    public UserHandler(String username, String channelOrigin) {
        this.username = username;
        this.channelOrigin = channelOrigin;
        userInventory = new Inventory(username, channelOrigin, this);

        if (Memebot.useMongo) {
            if (!Memebot.channelsPrivate.contains(this.channelOrigin)) {
                // todo old db code - remove soon
                //this.userCollection = Memebot.db.getCollection(this.channelOrigin + "_users");
                mongoHandler = new MongoHandler(Memebot.db, this.channelOrigin + "_users");
            } else {
                // todo old db code - remove soon
                //this.userCollection = Memebot.dbPrivate.getCollection(this.channelOrigin + "_users");
                mongoHandler = new MongoHandler(Memebot.dbPrivate, this.channelOrigin + "_users");
            }
        }

        readDB();
        if (dateJoined.equals("null")) {
            SimpleDateFormat sdfDate = new SimpleDateFormat("yyyy-MM-dd hh:mm:ss a"); // dd/MM/yyyy
            Calendar cal = Calendar.getInstance();
            this.dateJoined = sdfDate.format(cal.getTime());
        }

        setCommandPower(this.autoCommandPower);
    }

    public static Logger getLog() {
        return log;
    }

    public static void setLog(Logger log) {
        UserHandler.log = log;
    }

    private void readDB() {
        if (!Memebot.useMongo) {
            return;
        }

        try {
            mongoHandler.readDatabase(this.username);
        } catch (DatabaseReadException | IllegalArgumentException e) {
            log.warning(e.toString());
            this.newUser = true;
            return;
        }

        this.isModerator = (boolean) mongoHandler.getObject("mod", this.isModerator);
        this.points = (double) mongoHandler.getObject("pointsf", this.points);
        this.autogreet = mongoHandler.getObject("autogreet", this.autogreet).toString();
        this.customCommandPower = (int) mongoHandler.getObject("ccommandpower", this.customCommandPower);
        this.isUserBroadcaster = (boolean) mongoHandler.getObject("broadcaster", this.isUserBroadcaster);
        this.timeouts = (int) mongoHandler.getObject("timeouts", this.timeouts);
        this.enableAutogreets = (boolean) mongoHandler.getObject("enableautogreet", this.enableAutogreets);
        this.dateJoined = mongoHandler.getObject("datejoined", this.dateJoined).toString();
        this.timeStampJoined = (long) mongoHandler.getObject("timeStampJoined", this.timeStampJoined);
        this.nickname = mongoHandler.getObject("nickname", this.nickname).toString();
        this.walletSize = (double) mongoHandler.getObject("wallet", this.walletSize);
        this.isFollowing = (boolean) mongoHandler.getObject("isfollowing", this.isFollowing);
        this.hasFollowed = (boolean) mongoHandler.getObject("hasfollowed", this.hasFollowed);
        this.jackpotWins = (int) mongoHandler.getObject("jackpotwins", this.jackpotWins);
        this.lastTimeoutDuration = (int) mongoHandler.getObject("lasttoduration", this.lastTimeoutDuration);
        this.lastTimeoutReason = mongoHandler.getObject("lasttoreason", this.lastTimeoutReason).toString();

        userInventory.readDB();
    }

    public void setDB() {
        //Document userData = mongoHandler.getDocument();
        mongoHandler.updateDocument("_id", this.username);
        mongoHandler.updateDocument("pointsf", this.points);
        mongoHandler.updateDocument("mod", this.isModerator);
        mongoHandler.updateDocument("autogreet", this.autogreet);
        mongoHandler.updateDocument("ccommandpower", this.customCommandPower);
        mongoHandler.updateDocument("broadcaster", this.isUserBroadcaster);
        mongoHandler.updateDocument("timeouts", this.timeouts);
        mongoHandler.updateDocument("enableautogreet", this.enableAutogreets);
        mongoHandler.updateDocument("datejoined", this.dateJoined);
        mongoHandler.updateDocument("timeStampJoined", this.timeStampJoined);
        mongoHandler.updateDocument("nickname", this.nickname);
        mongoHandler.updateDocument("wallet", this.walletSize);
        mongoHandler.updateDocument("isfollowing", this.isFollowing);
        mongoHandler.updateDocument("hasfollowed", this.hasFollowed);
        mongoHandler.updateDocument("jackpotwins", this.jackpotWins);
        mongoHandler.updateDocument("lasttoduration", this.lastTimeoutDuration);
        mongoHandler.updateDocument("lasttoreason", this.lastTimeoutReason);
        //mongoHandler.setDocument(userData);
    }

    public void writeDB() {
        if (!Memebot.useMongo) {
            return;
        }
        setDB();

        mongoHandler.writeDatabase(this.username);

        userInventory.writeDB();
    }

    public void update(ChannelHandler channelHandler) {
        userInventory.update();

        // remove unused cooldowns asap
        for(String key : userCommandCooldowns.keySet()) {
            Cooldown cooldown = userCommandCooldowns.get(key);
            if(cooldown.canContinue()) {
                userCommandCooldowns.remove(key);
            }
        }
    }

    public boolean canRemove() {
        // remove user after 1 hour (0x36EE80 milliseconds) of inactivity
        HashMap<String, UserHandler> tmpUserList = new HashMap<String, UserHandler>();
        tmpUserList.put(this.username, this);

        //check if user has been marked for removal
        if (this.shouldBeRemoved && this.removeCooldown.canContinue()) {
            UserHandler.log.info("Removed user " + this.username);
            return true;
        }

        return false;
    }

    public void sendAutogreet(ChannelHandler channelHandler) {
        if (!this.hasAutogreeted && this.enableAutogreets && channelHandler.isAllowAutogreet() && !this.autogreet.equals("")) {
            channelHandler.sendMessage(autogreet, this.channelOrigin, this);
            this.hasAutogreeted = true;
        }
    }

    public String resetOAuth() {
        String keySource = username + Long.toString(timeStampJoined) + Integer.toHexString(new SecureRandom().nextInt());
        String oauth = Base64.encodeBase64String(keySource.getBytes());
        return oauth;
    }

    @Override
    public int compareTo(UserHandler another) {
        return username.compareTo(another.getUsername());
    }

    public MongoHandler getMongoHandler() {
        return mongoHandler;
    }

    public void setMongoHandler(MongoHandler mongoHandler) {
        this.mongoHandler = mongoHandler;
    }

    public int getJackpotWins() {
        return jackpotWins;
    }

    public void setJackpotWins(int jackpotWins) {
        this.jackpotWins = jackpotWins;
    }

    public String getUsername() {
        return username;
    }

    public void setUsername(String username) {
        this.username = username;
    }

    public String getChannelOrigin() {
        return channelOrigin;
    }

    public void setChannelOrigin(String channelOrigin) {
        this.channelOrigin = channelOrigin;
    }

    public boolean isModerator() {
        return isModerator;
    }

    public void setModerator(boolean moderator) {
        isModerator = moderator;
    }

    public boolean isUserBroadcaster() {
        return isUserBroadcaster;
    }

    public void setUserBroadcaster(boolean userBroadcaster) {
        isUserBroadcaster = userBroadcaster;
    }

    public boolean isNewUser() {
        return newUser;
    }

    public void setNewUser(boolean newUser) {
        this.newUser = newUser;
    }

    public String getNickname() {
        return nickname;
    }

    public void setNickname(String nickname) {
        this.nickname = nickname;
    }

    public int getCommandPower() {
        return commandPower;
    }

    public void setCommandPower(int commandPower) {
        this.autoCommandPower = commandPower;
        this.commandPower = commandPower + this.customCommandPower;
    }

    public int getAutoCommandPower() {
        return autoCommandPower;
    }

    public void setAutoCommandPower(int autoCommandPower) {
        this.autoCommandPower = autoCommandPower;
    }

    public int getCustomCommandPower() {
        return customCommandPower;
    }

    public void setCustomCommandPower(int customCommandPower) {
        this.customCommandPower = customCommandPower;
    }

    public double getPoints() {
        return points;
    }

    public boolean setPoints(double f) {
        boolean result = true;

        this.points = f;
        if (this.points < 0) {
            this.points = 0;
        }

        for (ChannelHandler ch : Memebot.joinedChannels) {
            if (ch.getChannel().equals(this.channelOrigin)) {
                if (this.points + f > ch.getMaxPoints() || (this.points + f > this.walletSize && this.walletSize > 0)) {
                    if (Memebot.debug) {
                        this.points = ch.getMaxPoints();
                        result = false;
                    } else {
                        result = true;
                    }
                }
            }
        }

        this.writeDB();

        return result;
    }

    public Cooldown getUserCooldown() {
        return userCooldown;
    }

    public void setUserCooldown(Cooldown userCooldown) {
        this.userCooldown = userCooldown;
    }

    public String getAutogreet() {
        return autogreet;
    }

    public void setAutogreet(String autogreet) {
        this.autogreet = autogreet;
    }

    public int getTimeouts() {
        return timeouts;
    }

    public void setTimeouts(int timeouts) {
        this.timeouts = timeouts;
    }

    public HashMap<String, Cooldown> getUserCommandCooldowns() {
        return userCommandCooldowns;
    }

    public void setUserCommandCooldowns(HashMap<String, Cooldown> userCommandCooldowns) {
        this.userCommandCooldowns = userCommandCooldowns;
    }

    public String getModNote() {
        return modNote;
    }

    public void setModNote(String modNote) {
        this.modNote = modNote;
    }

    public SecureRandom getRandom() {
        return random;
    }

    public void setRandom(SecureRandom random) {
        this.random = random;
    }

    public String getDateJoined() {
        return dateJoined;
    }

    public void setDateJoined(String dateJoined) {
        this.dateJoined = dateJoined;
    }

    public long getTimeStampJoined() {
        return timeStampJoined;
    }

    public void setTimeStampJoined(long timeStampJoined) {
        this.timeStampJoined = timeStampJoined;
    }

    public boolean isEnableAutogreets() {
        return enableAutogreets;
    }

    public void setEnableAutogreets(boolean enableAutogreets) {
        this.enableAutogreets = enableAutogreets;
    }

    public double getWalletSize() {
        return walletSize;
    }

    public void setWalletSize(double walletSize) {
        this.walletSize = walletSize;
    }

    public boolean isFollowing() {
        return isFollowing;
    }

    public void setFollowing(boolean following) {
        isFollowing = following;
    }

    public boolean isHasFollowed() {
        return hasFollowed;
    }

    public void setHasFollowed(boolean hasFollowed) {
        this.hasFollowed = hasFollowed;
    }

    public boolean isHasAutogreeted() {
        return hasAutogreeted;
    }

    public void setHasAutogreeted(boolean hasAutogreeted) {
        this.hasAutogreeted = hasAutogreeted;
    }

    public Cooldown getRemoveCooldown() {
        return removeCooldown;
    }

    public void setRemoveCooldown(Cooldown removeCooldown) {
        this.removeCooldown = removeCooldown;
    }

    public boolean isShouldBeRemoved() {
        return shouldBeRemoved;
    }

    public void setShouldBeRemoved(boolean shouldBeRemoved) {
        this.shouldBeRemoved = shouldBeRemoved;
    }

    public Inventory getUserInventory() {
        return userInventory;
    }

    public void setUserInventory(Inventory userInventory) {
        this.userInventory = userInventory;
    }

    public int getLastTimeoutDuration() {
        return lastTimeoutDuration;
    }

    public void setLastTimeoutDuration(int lastTimeoutDuration) {
        this.lastTimeoutDuration = lastTimeoutDuration;
    }

    public String getLastTimeoutReason() {
        return lastTimeoutReason;
    }

    public void setLastTimeoutReason(String lastTimeoutReason) {
        this.lastTimeoutReason = lastTimeoutReason;
    }

    public String screenName() {
        if (!this.nickname.isEmpty()) {
            return nickname;
        }
        return this.username;
    }

    public String getOauth() {
        String oauth = resetOAuth();
        MongoHandler mongoHandler = new MongoHandler(Memebot.db, "#oauth#");
        try {
            mongoHandler.readDatabase(username);
        } catch (DatabaseReadException e) {
            log.info(e.toString());
            setOauth(oauth);
            return oauth;
        }

        String dbOauth = mongoHandler.getDocument().getOrDefault("oauth", oauth).toString();

        if (oauth.equals(dbOauth)) {
            setOauth(dbOauth);
        }

        return dbOauth;
    }

    public void setOauth(String oauth) {
        MongoHandler mongoHandler = new MongoHandler(Memebot.db, "#oauth#");

        Document document = new Document();
        document.append("oauth", oauth).append("_id", username);
        mongoHandler.setDocument(document);
        mongoHandler.writeDatabase(username);
    }

    public String getAPIKey() {
        String apikey = resetOAuth();
        MongoHandler mongoHandler = new MongoHandler(Memebot.db, "#apikey#");
        try {
            mongoHandler.readDatabase(username);
        } catch (DatabaseReadException e) {
            log.info(e.toString());
            setAPIKey(apikey);
            return apikey;
        }

        String dbkey = mongoHandler.getDocument().getOrDefault("apikey", apikey).toString();

        if (apikey.equals(dbkey)) {
            setOauth(dbkey);
        }

        return dbkey;
    }

    public void setAPIKey(String key) {
        MongoHandler mongoHandler = new MongoHandler(Memebot.db, "#apikey#");

        Document document = new Document();
        document.append("apikey", key).append("_id", username);
        mongoHandler.setDocument(document);
        mongoHandler.writeDatabase(username);
    }

    public JSONObject toJSONObject() {
        JSONObject jsonObject = new JSONObject();

        jsonObject.put("points", points);
        jsonObject.put("timeouts", timeouts);
        jsonObject.put("_id", username);
        jsonObject.put("_channel", channelOrigin);
        jsonObject.put("_self", Memebot.webBaseURL + "/api/users/" + channelOrigin.replace("#", "") + "/" + username);
        jsonObject.put("joinded_t", timeStampJoined);
        jsonObject.put("joined_str", dateJoined);
        jsonObject.put("inventory", userInventory.toJSON());

        return jsonObject;
    }

    public String toJSON() {
        return toJSONObject().toJSONString();
    }
}
